File: APP03022-mrch.backend.somx.catalogos-api/src/main/java/com/sodimac/catman/api/mapper/CatalogElementMapper.java
package com.sodimac.catman.api.mapper;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Component;

import com.sodimac.catman.api.model.dto.CatalogElementDto;
import com.sodimac.catman.api.model.dto.CatalogSimpleDto;
import com.sodimac.catman.api.model.entity.CatalogDetail;
import com.sodimac.catman.api.model.entity.CatalogHeader;
import com.sodimac.catman.api.model.entity.DictionaryLang;
import com.sodimac.catman.api.repository.CatalogDetailRepository;
import com.sodimac.catman.api.repository.CatalogHeaderRepository;
import com.sodimac.catman.api.repository.DictionaryLangRepository;

@Component
public class CatalogElementMapper {

    private static final int DEFAULT_LANG_ID = 1;

    private final CatalogHeaderRepository headerRepository;
    private final CatalogDetailRepository detailRepository;
    private final DictionaryLangRepository dictionaryLangRepository;

    public CatalogElementMapper(CatalogHeaderRepository headerRepository,
                                CatalogDetailRepository detailRepository,
                                DictionaryLangRepository dictionaryLangRepository) {
        this.headerRepository = headerRepository;
        this.detailRepository = detailRepository;
        this.dictionaryLangRepository = dictionaryLangRepository;
    }

    public CatalogElementDto toDto(CatalogDetail entity) {
        if (entity == null) {
            return null;
        }

        String elementName = entity.getKey();
        if (entity.getDictId() != null && entity.getDictId() > 0) {
            elementName = dictionaryLangRepository.findByDictIdAndLangId(entity.getDictId(), DEFAULT_LANG_ID)
                    .map(DictionaryLang::getDescription)
                    .orElse(entity.getKey());
        }

        CatalogElementDto dto = CatalogElementDto.builder()
                .id(entity.getId())
                .catalogId(entity.getHeader() != null ? entity.getHeader().getId() : null)
                .catalogCode(entity.getHeader() != null ? entity.getHeader().getCode() : null)
                .element(elementName)
                .value(entity.getValue())
                .key(entity.getKey())
                .validFrom(entity.getValidFrom())
                .validTo(entity.getValidTo())
                .status(entity.getStatus())
                .statusDescription(entity.getStatus() == 1 ? "Activo" : "Inactivo")
                .parentCatalogId(entity.getParentCatalogId())
                .parentElementId(entity.getParentElementId())
                .createdBy(entity.getCreatedBy())
                .createdAt(entity.getCreatedAt())
                .updatedBy(entity.getUpdatedBy())
                .updatedAt(entity.getUpdatedAt())
                .externalKey(entity.getExternalKey())
                .sortOrder(entity.getSortOrder())
                .attributes(entity.getAttributes())
                .build();

        if (entity.getParentCatalogId() != null) {
            headerRepository.findById(entity.getParentCatalogId())
                    .ifPresent(parent -> dto.setParentCatalogName(parent.getName()));
        }

        if (entity.getParentElementId() != null) {
            detailRepository.findById(entity.getParentElementId())
                    .ifPresent(parent -> {
                        String parentName = parent.getKey();
                        if (parent.getDictId() != null && parent.getDictId() > 0) {
                            parentName = dictionaryLangRepository.findByDictIdAndLangId(parent.getDictId(), DEFAULT_LANG_ID)
                                    .map(DictionaryLang::getDescription)
                                    .orElse(parent.getKey());
                        }
                        dto.setParentElementName(parentName);
                    });
        }

        return dto;
    }

    public List<CatalogElementDto> toDtoList(List<CatalogDetail> entities) {
        return entities.stream()
                .map(this::toDto)
                .collect(Collectors.toList());
    }

    public CatalogSimpleDto toSimpleDto(CatalogHeader entity) {
        if (entity == null) {
            return null;
        }

        return CatalogSimpleDto.builder()
                .id(entity.getId())
                .code(entity.getCode())
                .name(entity.getName())
                .description(entity.getDescription())
                .catalogType(entity.getCatalogType())
                .status(entity.getStatus())
                .build();
    }

    public List<CatalogSimpleDto> toSimpleDtoList(List<CatalogHeader> entities) {
        return entities.stream()
                .map(this::toSimpleDto)
                .collect(Collectors.toList());
    }
}

File: APP03022-mrch.backend.somx.util-api/src/mappers/catalogElement.mapper.ts
import type { CatalogDetail } from '@/entities/CatalogDetail.entity.js';
import type { CatalogHeader } from '@/entities/CatalogHeader.entity.js';
import type { CatalogElementDto, CatalogSimpleDto } from '@/dto/catalog.dto.js';
import * as headerRepo from '@/repositories/catalogHeader.repo.js';
import * as detailRepo from '@/repositories/catalogDetail.repo.js';
import * as dictRepo from '@/repositories/dictionaryLang.repo.js';

const DEFAULT_LANG_ID = 1;

async function resolveElementName(entity: CatalogDetail): Promise<string> {
    if (entity.dictId == null || entity.dictId <= 0) return entity.key;
    const dict = await dictRepo.findByDictIdAndLangId(entity.dictId, DEFAULT_LANG_ID);
    return dict?.description ?? entity.key;
}

function buildBaseDto(entity: CatalogDetail, elementName: string): CatalogElementDto {
    return {
        id: entity.id,
        catalogId: entity.header?.id ?? entity.headerId ?? null,
        catalogCode: entity.header?.code ?? null,
        element: elementName,
        value: entity.value ?? null,
        key: entity.key,
        validFrom: entity.validFrom ?? null,
        validTo: entity.validTo ?? null,
        status: entity.status ?? null,
        statusDescription: entity.status === 1 ? 'Activo' : 'Inactivo',
        parentCatalogId: entity.parentCatalogId ?? null,
        parentElementId: entity.parentElementId ?? null,
        createdBy: entity.createdBy ?? null,
        createdAt: entity.createdAt ?? null,
        updatedBy: entity.updatedBy ?? null,
        updatedAt: entity.updatedAt ?? null,
        externalKey: entity.externalKey ?? null,
        sortOrder: entity.sortOrder ?? null,
        attributes: entity.attributes ?? null,
    };
}

async function resolveParentCatalogName(parentCatalogId: number | null | undefined): Promise<string | undefined> {
    if (parentCatalogId == null) return undefined;
    const parent = await headerRepo.findById(parentCatalogId);
    return parent?.name;
}

async function resolveParentElementName(parentElementId: number | null | undefined): Promise<string | undefined> {
    if (parentElementId == null) return undefined;
    const parent = await detailRepo.findById(parentElementId);
    if (!parent) return undefined;
    if (parent.dictId == null || parent.dictId <= 0) return parent.key;
    const pdict = await dictRepo.findByDictIdAndLangId(parent.dictId, DEFAULT_LANG_ID);
    return pdict?.description ?? parent.key;
}

export async function toDto(entity: CatalogDetail | null | undefined): Promise<CatalogElementDto | null> {
    if (!entity) return null;

    const elementName = await resolveElementName(entity);
    const dto = buildBaseDto(entity, elementName);

    const parentCatalogName = await resolveParentCatalogName(entity.parentCatalogId);
    if (parentCatalogName !== undefined) {
        dto.parentCatalogName = parentCatalogName;
    }

    const parentElementName = await resolveParentElementName(entity.parentElementId);
    if (parentElementName !== undefined) {
        dto.parentElementName = parentElementName;
    }

    return dto;
}

export async function toDtoList(entities: CatalogDetail[]): Promise<CatalogElementDto[]> {
    const results = await Promise.all(entities.map(e => toDto(e)));
    return results.filter((d): d is CatalogElementDto => d !== null);
}

export function toSimpleDto(entity: CatalogHeader | null | undefined): CatalogSimpleDto | null {
    if (!entity) return null;
    return {
        id: entity.id,
        code: entity.code,
        name: entity.name,
        description: entity.description ?? null,
        catalogType: entity.catalogType ?? null,
        status: entity.status
    };
}

export function toSimpleDtoList(entities: CatalogHeader[]): CatalogSimpleDto[] {
    return entities
        .map(e => toSimpleDto(e))
        .filter((d): d is CatalogSimpleDto => d !== null);
}

